File: my-folder/0211-design-add-and-search-words-data-structure/solution.py
class WordDictionary:

    def __init__(self):
        self.children = [None] * 26
        self.check = False


    def addWord(self, word: str) -> None:
        curr = self
        for letter in word:
            if not curr.children[ord(letter) - ord('a')]:
                curr.children[ord(letter) - ord('a')] = WordDictionary()
            curr = curr.children[ord(letter) - ord('a')]
        curr.check = True


    def search(self, word: str) -> bool:
        curr = self
        for i in range(len(word)):
            char = word[i]
            if char == '.':
                for child in curr.children:
                    if child and child.search(word[i+1:]):
                        return True
                return False
            if not curr.children[ord(char) - ord('a')]:
                return False
            curr = curr.children[ord(char) - ord('a')]

        return curr and curr.check

# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)

File: my-folder/0211-design-add-and-search-words-data-structure/solution.java
class WordDictionary {

    WordDictionary[] children;
    boolean check;

    public WordDictionary() {
        children = new WordDictionary[26];
        check = false;
    }
    
    public void addWord(String word) {
        WordDictionary current = this;
        for (char c : word.toCharArray()) {
            if (current.children[c - 'a'] == null) {
                current.children[c - 'a'] = new WordDictionary();
            }
            current = current.children[c - 'a'];
        }
        current.check = true;
    }
    
    public boolean search(String word) {
        WordDictionary current = this;
        for (int i = 0; i < word.length(); i++) {
            char c = word.charAt(i);
            if (c == '.') {
                for (WordDictionary child : current.children) {
                    if (child != null && child.search(word.substring(i+1, word.length()))) {
                        return true;
                    }
                }
                return false;
            }
            else {
                if (current.children[c - 'a'] == null) {
                    return false;
                }
            }
            current = current.children[c - 'a'];
        }
        return current != null && current.check;
    }
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * boolean param_2 = obj.search(word);
 */
